File: web/src/features/identity/keys/kid.ts
/**
 * Key identifier (kid) derivation
 * Matches backend implementation: SHA256(pubkey) encoded as base64url
 */

import { sha256 } from '@noble/hashes/sha2.js';
import { encodeBase64Url } from './utils';

/**
 * Derive a key identifier (kid) from a public key.
 * Uses SHA-256 hash of the public key, encoded as base64url.
 *
 * @param publicKey - Ed25519 public key (32 bytes)
 * @returns Base64url-encoded SHA-256 hash
 */
export function deriveKid(publicKey: Uint8Array): string {
  const hash = sha256(publicKey);
  return encodeBase64Url(hash);
}

File: service/src/identity/crypto/kid.rs
use base64::engine::general_purpose::URL_SAFE_NO_PAD;
use base64::Engine;
use sha2::{Digest, Sha256};

/// Derive a key identifier (kid) from a public key.
/// KID = base64url(SHA-256(pubkey)[0:16])
///
/// # Must Use
/// Always consume the returned string; dropping it will skip caller validation.
#[must_use]
pub fn derive_kid(public_key: &[u8]) -> String {
    let hash = Sha256::digest(public_key);
    // Truncate to first 16 bytes for shorter KIDs (matches frontend)
    URL_SAFE_NO_PAD.encode(&hash[..16])
}
